convex/aiTown/player.ts

```diff
@@ -0,0 +1,305 @@
+    }
+
+    // Stop pathfinding if we've reached our destination.
+    if (pathfinding.state.kind === 'moving' && pointsEqual(pathfinding.destination, position)) {
+      stopPlayer(this);
+    }
+
+    // Stop pathfinding if we've timed out.
+    if (pathfinding.started + PATHFINDING_TIMEOUT < now) {
+      console.warn(`Timing out pathfinding for ${this.id}`);
+      stopPlayer(this);
+    }
+
+    // Transition from "waiting" to "needsPath" if we're past the deadline.
+    if (pathfinding.state.kind === 'waiting' && pathfinding.state.until < now) {
+      pathfinding.state = { kind: 'needsPath' };
+    }
+
+    // Perform pathfinding if needed.
+    if (pathfinding.state.kind === 'needsPath') {
+      const route = findRoute(game, now, this, pathfinding.destination);
+      if (route === null) {
+        console.log(`Failed to route to ${JSON.stringify(pathfinding.destination)}`);
+        stopPlayer(this);
+      } else {
+        if (route.newDestination) {
+          console.warn(
+            `Updating destination from ${JSON.stringify(
+              pathfinding.destination,
+            )} to ${JSON.stringify(route.newDestination)}`,
+          );
+          pathfinding.destination = route.newDestination;
+        }
+        pathfinding.state = { kind: 'moving', path: route.path };
+      }
+    }
+  }
+
+  tickPosition(game: Game, now: number) {
+    // There's nothing to do if we're not moving.
+    if (!this.pathfinding || this.pathfinding.state.kind !== 'moving') {
+      this.speed = 0;
+      return;
+    }
+
+    // Compute a candidate new position and check if it collides
+    // with anything.
+    const candidate = pathPosition(this.pathfinding.state.path as any, now);
+    if (!candidate) {
+      console.warn(`Path out of range of ${now} for ${this.id}`);
+      return;
+    }
+    const { position, facing, velocity } = candidate;
+    const collisionReason = blocked(game, now, position, this.id);
+    if (collisionReason !== null) {
+      const backoff = Math.random() * PATHFINDING_BACKOFF;
+      console.warn(`Stopping path for ${this.id}, waiting for ${backoff}ms: ${collisionReason}`);
+      this.pathfinding.state = {
+        kind: 'waiting',
+        until: now + backoff,
+      };
+      return;
+    }
+    // Update the player's location.
+    this.position = position;
+    this.facing = facing;
+    this.speed = velocity;
+  }
+
+  static join(
+    game: Game,
+    now: number,
+    name: string,
+    character: string,
+    description: string,
+    tokenIdentifier?: string,
+  ) {
+    if (tokenIdentifier) {
+      let numHumans = 0;
+      for (const player of game.world.players.values()) {
+        if (player.human) {
+          numHumans++;
+        }
+        if (player.human === tokenIdentifier) {
+          throw new Error(`You are already in this game!`);
+        }
+      }
+      if (numHumans >= MAX_HUMAN_PLAYERS) {
+        throw new Error(`Only ${MAX_HUMAN_PLAYERS} human players allowed at once.`);
```

Comment:
will this propagate all the way back via useSendInput? I'm guessing error messages get translated into the input result but haven't read that yet

Reply:
yep! manually tested it too